src/compiler/checker.ts

```diff
@@ -11143,12 +11144,13 @@ namespace ts {
                     someHaveQuestionToken = someHaveQuestionToken || hasQuestionToken(node);
                     allHaveQuestionToken = allHaveQuestionToken && hasQuestionToken(node);
 
-                    if (nodeIsPresent(node.body) && bodyDeclaration) {
+                    if (nodeIsPresent(node.body) && implementationDeclaration) {
                         if (isConstructor) {
                             multipleConstructorImplementation = true;
                         }
                         else {
                             duplicateFunctionDeclaration = true;
+                            duplicateDefaultExports = !!(node.flags & NodeFlags.Default);

```

Comment:
At this point we're not sure that it's a duplicate default export, right? Maybe the name should be different.
(I could be wrong, I've only skimmed `areAllFunctionDeclarations...` so far)